client/state/data-layer/wpcom-http/index.js

```diff
@@ -0,0 +1,60 @@
+/**
+ * External dependencies
+ */
+import { get, noop } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import wpcom from 'lib/wp';
+import { WPCOM_HTTP_REQUEST } from 'state/action-types';
+import { extendAction } from 'state/utils';
+
+/**
+ * Returns the appropriate fetcher in wpcom given the request method
+ *
+ * fetcherMap :: String -> (Params -> Query -> [Body] -> Promise)
+ *
+ * @param {string} method name of HTTP method for request
+ * @returns {Function} the fetcher
+ */
+const fetcherMap = method => get( {
+	GET: wpcom.req.get.bind( wpcom.req ),
+	POST: wpcom.req.post.bind( wpcom.req ),
+}, method, noop );
+
+export const successMeta = data => ( { meta: { dataLayer: { data } } } );
+export const failureMeta = error => ( { meta: { dataLayer: { error } } } );
+export const progressMeta = ( { size: total, loaded } ) => ( { meta: { dataLayer: { progress: { total, loaded } } } } );
+
+const queueRequest = ( { dispatch }, action, next ) => {
+	const {
+		body,
+		formData,
+		method,
+		onSuccess,
+		onFailure,
+		onProgress,
+		path,
+		query,
+	} = action;
+
+	const request = fetcherMap( method.toUpperCase() )(
+		{ path, formData },
```

Comment:
it started out as a generic HTTP layer and somehow it turned into a WPCOM layer.

@aduth I think it's better as a generic handler and mainly it morphed because of the `path` variable. the better approach is probably to make one generic HTTP middleware and then create a WordPress.com API action creator. when I move next on this PR I will try and separate those out

```js
// create generic HTTP request
dispatch( http( {
	method: 'GET',
	url: 'http://api.icndb.com/jokes/random',
	onSuccess: { type: JOKE_ADD }
} ) )

dispatch( wpcomApi( {
	method: 'GET',
	path: '/me',
	onSuccess: action,
	onFailure: action,
} ) )
```